src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Completion/RemoteCompletionService.cs

```diff
@@ -0,0 +1,229 @@
+                ProvisionalTextEdit = provisionalCompletionInfo.ProvisionalTextEdit
+            };
+        }
+
+        var shouldIncludeSnippets = positionInfo.LanguageKind == RazorLanguageKind.Html
+            && await DelegatedCompletionHelper.ShouldIncludeSnippetsAsync(remoteDocumentContext, index, cancellationToken);
+
+        return new CompletionPositionInfo() { DocumentPositionInfo = positionInfo, ShouldIncludeDelegationSnippets = shouldIncludeSnippets };
+    }
+
+    public ValueTask<Response> GetCompletionAsync(
+        JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo,
+        JsonSerializableDocumentId documentId,
+        CompletionPositionInfo positionInfo,
+        VSInternalCompletionContext completionContext,
+        VSInternalClientCapabilities clientCapabilities,
+        RazorCompletionOptions razorCompletionOptions,
+        HashSet<string> existingHtmlCompletions,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetCompletionAsync(
+                context,
+                positionInfo,
+                completionContext,
+                clientCapabilities,
+                razorCompletionOptions,
+                existingHtmlCompletions,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> GetCompletionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        CompletionPositionInfo positionInfo,
+        VSInternalCompletionContext completionContext,
+        VSInternalClientCapabilities clientCapabilities,
+        RazorCompletionOptions razorCompletionOptions,
+        HashSet<string> existingDelegatedCompletions,
+        CancellationToken cancellationToken)
+    {
+        VSInternalCompletionList? csharpCompletionList = null;
+        var documentPositionInfo = positionInfo.DocumentPositionInfo;
+        if (documentPositionInfo.LanguageKind == RazorLanguageKind.CSharp &&
+            CompletionTriggerCharacters.IsValidTrigger(CompletionTriggerCharacters.CSharpTriggerCharacters, completionContext))
+        {
+            var mappedPosition = documentPositionInfo.Position;
+            csharpCompletionList = await GetCSharpCompletionAsync(
+                    remoteDocumentContext,
+                    documentPositionInfo.HostDocumentIndex,
+                    mappedPosition,
+                    positionInfo.ProvisionalTextEdit,
+                    completionContext,
+                    clientCapabilities,
+                    razorCompletionOptions,
+                    cancellationToken);
+
+            if (csharpCompletionList is not null)
+            {
+                Debug.Assert(existingDelegatedCompletions.Count == 0, "Delegated completion should be either C# or HTML, not both");
+                existingDelegatedCompletions.UnionWith(csharpCompletionList.Items.Select((item) => item.Label));
```

Comment:
I haven't looked too much into the code using it - I just know that's what the current code does. We always get labels from delegated completion items, whether C# or HTML. See https://github.com/dotnet/razor/blob/a91bdccb34afbb22701ffb7fff9134046e532c62/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/CompletionListProvider.cs#L47C13-L47C27 - not to say it was doing the right thing :)

Reply:
That is surprising, but it looks like they were only used in tag helper completion, which would only have occurred in a Html context. At least I would hope so anyway. I guess since this list isn't being serialized anywhere it's fine to leave, but I would think a lot of this code gets much easier to reason about if we simplify it at some point.

Reply:
I'll leave it for now since that's what the current code does, but I did file [https://github.com/dotnet/razor/issues/11103](https://github.com/dotnet/razor/issues/11103) to investigate further.